server/src/main/java/org/elasticsearch/cluster/routing/RelocationFailureInfo.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.routing;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.xcontent.ToXContentFragment;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Holds additional information as to why the shard failed to relocate.
+ */
+public class RelocationFailureInfo implements ToXContentFragment, Writeable {
```

Comment:
Converted to a record. I wanted to keep it a class/record in case we want to make this behavior more complex in the future (similar to UnassignedInfo with recording failed nodes or introducing a delay)